host/s3.go

```diff
@@ -0,0 +1,412 @@
+	)
+	opCtx, cancel := context.WithTimeout(ctx, S3ConnectionTimeout)
+	defer cancel()
+
+	err := s3ds.testConnection(opCtx)
+	if err != nil {
+		return err
+	}
+
+	s3ds.spawnPeriodicFlusher(ctx, s3ds.config.FlushInterval)
+
+	mainCtx, mainCancel := context.WithCancel(ctx)
+	s3ds.ctx = mainCtx
+	s3ds.cancelFn = mainCancel
+
+	<-mainCtx.Done()
+	return mainCtx.Err()
+}
+
+func (s3ds *S3DataStream) Stop(ctx context.Context) error {
+	// stop the mainCtx, as we don't want new traces coming in
+	s3ds.cancelFn()
+	// ensure to prune existing traces before closing
+	opCtx, cancel := context.WithTimeout(ctx, S3OpTimeout)
+	defer cancel()
+	s3ds.submitRecords(opCtx)
+	return nil
+}
+
+func (s3ds *S3DataStream) PutRecord(ctx context.Context, event *TraceEvent) error {
+	s3ds.batcher.addNewTrace(event)
+	if s3ds.batcher.isFull() {
+		s3ds.submitRecords(ctx)
+	}
+	return nil
+}
+
+// submitRecords is the private method that will:
+// 1. get the copy of the existing traces, leaving a new batcher empty
+// 2. create the necessary s3 key to submite the batched traces
+// 3. format the traces into a parquet file
+// 4. push the file into the s3 bucket
+// *NOTE*: the method leaves a routine in the background to format the parquet
+//
+//	file, and to submit it into S3. Reason -> to avoid freezing other processes
+func (s3ds *S3DataStream) submitRecords(ctx context.Context) {
+	// check if the trace-buffer has anything
+	if s3ds.batcher.len() <= 0 {
+		return
+	}
+
+	// get and reset the traces from the batcher
+	// increase the counter for the next file descriptor
+	// and get the peer_id from the traces
+	traces := s3ds.batcher.reset()
+	currentFileCnt := s3ds.fileCnt.Load()
+	s3ds.fileCnt.Add(1)
+	producerID := traces[0].PeerID
+
+	// compose the path for the s3 key/file
+	// s3Path = hermes/peer_id/year/month/day/hour/file_index.parquet
+	t := time.Now()
+	s3Key := fmt.Sprintf(
+		"hermes/%s/%d/%d/%d/%d/%d.parquet",
+		producerID,
+		t.Year(),
+		t.Month(),
+		t.Day(),
+		t.Hour(),
+		currentFileCnt,
+	)
+	slog.Info(
+		"submitting traces to s3",
+		slog.Attr{Key: "traces", Value: slog.AnyValue(len(traces))},
+	)
+
+	go func() {
+		// write the traces into the s3 file descriptor
+		parquetBuffer := new(bytes.Buffer)
+		schema := parquet.SchemaOf(new(ParquetTraceEvent))
+		pw := parquet.NewGenericWriter[ParquetTraceEvent](parquetBuffer, schema)
+
+		_, err := pw.Write(traces)
+		if err != nil {
+			slog.Error(
+				"writing traces to parquet",
+				tele.LogAttrError(err))
+		}
+		err = pw.Close()
+		if err != nil {
+			slog.Error("unable to close the parquer writer", tele.LogAttrError(err))
+			return
+		}
+
+		// submit the resulting bytes
+		err = s3ds.s3KeySubmission(ctx, s3Key, parquetBuffer.Bytes())
+		if err != nil {
+			slog.Error("uploading file to s3", tele.LogAttrError(err))
+		} else {
+			slog.Info(
+				"submitted file to s3",
+				slog.Attr{Key: "bytes", Value: slog.IntValue(len(parquetBuffer.Bytes()))},
+			)
+		}
+	}()
```

Comment:
This is an unbounded go routine. I would add a worker pool of let's say 10 workers that submit traces (doing it in parallel might influence event ordering?). If the workers cannot keep up with data ingress we can at least log a message. At least we have made sure that Hermes won't fall apart.

Reply:
this should be resolved on the lastest commits